axelor-production/src/main/java/com/axelor/apps/production/service/SaleOrderLineBomServiceImpl.java

```diff
@@ -45,4 +68,168 @@ public List<SaleOrderLine> createSaleOrderLinesFromBom(
+      var isInSubList =
+          saleOrderLine.getSubSaleOrderLineList().stream()
+              .map(SaleOrderLine::getBillOfMaterialLine)
+              .filter(Objects::nonNull)
+              .anyMatch(bomLine -> bomLine.equals(billOfMaterialLine));
+
+      logger.debug(
+          "Checking existence of billOfMaterialLine {} in {}",
+          billOfMaterialLine,
+          saleOrderLine.getSubSaleOrderLineList());
+      if (!isInSubList) {
+        logger.debug("BomLine does not exist, removing it");
+        bom.removeBillOfMaterialLineListItem(billOfMaterialLine);
+      }
+    }
+    billOfMaterialRepository.save(bom);
+    logger.debug("Updated saleOrderLine {} with bom {}", saleOrderLine, bom);
+  }
+
+  protected BillOfMaterialLine createBomLineFrom(SaleOrderLine subSaleOrderLine) {
+    return billOfMaterialLineService.createBillOfMaterialLine(
+        subSaleOrderLine.getProduct(),
+        subSaleOrderLine.getBillOfMaterial(),
+        subSaleOrderLine.getQty(),
+        subSaleOrderLine.getUnit(),
+        0,
+        subSaleOrderLine.getProduct().getStockManaged());
+  }
+
+  @Transactional(rollbackOn = Exception.class)
+  protected BillOfMaterial customizeBomOf(SaleOrderLine saleOrderLine, int depth)
+      throws AxelorException {
+    if (saleOrderLine.getBillOfMaterial() == null) {
+      return null;
+    }
+    if (depth > 1000) {
+      throw new AxelorException(
+          TraceBackRepository.CATEGORY_INCONSISTENCY,
+          I18n.get(ProductionExceptionMessage.MAX_DEPTH_REACHED));
+    }
+
+    var billOfMaterial = saleOrderLine.getBillOfMaterial();
+    var noOfPersonalizedBOM =
+        billOfMaterialRepository
+                .all()
+                .filter(
+                    "self.product = ?1 AND self.personalized = true", billOfMaterial.getProduct())
+                .count()
+            + 1;
+    var personalizedBOM = JPA.copy(billOfMaterial, false);
+    var name =
+        personalizedBOM.getName()
+            + " ("
+            + I18n.get(ProductionExceptionMessage.BOM_1)
+            + " "
+            + noOfPersonalizedBOM
+            + ")";
+    personalizedBOM.setName(name);
+    personalizedBOM.setPersonalized(true);
+    saleOrderLine.setBillOfMaterial(personalizedBOM);
+
+    if (saleOrderLine.getSubSaleOrderLineList() != null) {
+      for (SaleOrderLine subSaleOrderLine : saleOrderLine.getSubSaleOrderLineList()) {
+        var bomLine = createBomLineFrom(subSaleOrderLine);
+        // If it is not personalized, we will customize, else just use the personalized one.
+        if (subSaleOrderLine.getIsToProduce() && !bomLine.getBillOfMaterial().getPersonalized()) {
+          subSaleOrderLine.setBillOfMaterial(customizeBomOf(subSaleOrderLine, depth + 1));
+        }
+        // Relink billOfMaterialLine
+        subSaleOrderLine.setBillOfMaterialLine(bomLine);
+        personalizedBOM.addBillOfMaterialLineListItem(bomLine);
+      }
+    }
+
```

Comment:
it looks like part of this is copy pasted from `BillOfMaterialService#customizeBillOfMaterial`, the service should be called instead.